### DISCUSSION TITLE
Use kelvins as the preferred white temperature unit

### DISCUSSION THREAD

**Discussion by amelchio** [home-assistant/architecture#564] [Entity Models] [answered]
Home Assistant currently uses mireds as the internal representation of color temperature for white lights. I believe this has no particular reasoning other than the Hue API being the first one implemented and it uses mireds.

However, lights are always advertised with a certain kelvin range, even Hue bulbs, so this is the scale that humans know and also the one used for most other light APIs.

This means that almost all light integrations must do kelvin/mired conversions and that has resulted in numerous bugs. We have had bugs such as swapped warm/cold, division by zero and setting a temperature that is slightly (but noticeably) wrong. These kinds of bugs continue to appear in new implementations. This is because implementors expect kelvins and mireds to be related by a linear function but they are not.

Since the kelvin/mired relationship is reciprocal rather than linear, an interesting artifact is that the color temperature slider in the UI is not intuitive. For a 2000K-8000K (500-125 mireds) bulb, the expected midpoint is 5000K but the midpoint we currently use in the UI is way off at 3200K.

# Proposal

* We currently have ATTR_KELVIN as an alternative way to provide ATTR_COLOR_TEMP. I propose to reverse this relationship, making ATTR_KELVIN the primary attribute that platforms receive.
* Introduce a `kelvin` property for the color temperature of a light. If needed, we can keep `color_temp` centrally, just like `hs_color` auto-converts to `rgb_color`.
* Attributes `min_mireds`/`max_mireds` become `min_kelvin`/`max_kelvin` or maybe today we have a better way to provide these values to the frontend?
* Rename `SUPPORT_COLOR_TEMP` to `SUPPORT_KELVIN`
* Change the UI to use a kelvin scale.

If we can agree, I volunteer to implement these changes across all light platforms. It is not entirely clear to me how it would affect external stuff like mqtt?

**Comment by OttoWinter** [comment, marked as answer]
I 100% agree with this. The internal use of mired has always annoyed me very much (without HA I probably wouldn't even have had to learn what that scale is)

**Reply by bdraco** [reply]
This comes up over and over and over again.  This is a good proposal.  We should make this change.

**Comment by amelchio** [comment]
I am closing this issue. It never received much discussion and though I see that it got a fair number of 👍, I have now lost the motivation to go forward with it.

**Comment by james10952001** [comment]
Perhaps we could revisit this at some point? There are some reasonable technical reasons for using the rather esoteric Mireds unit internally however the whole lighting industry uses the Kelvin scale so it would be really great the UI could be set to Kelvin instead. This is similar to C vs F for temperature, it isn't THAT difficult to learn to work with a new scale but it's a lot more convenient to be able to use the one most people are accustomed to.

**Comment by HenkRenting** [comment]
I fully agree.
Kelvin should be used instead of Mired.
I never even heard of it.. :-)

**Comment by icemanch** [comment]
Just adding to this.  Kelvin is the standard for describing light color.  People understand Kelvin.  Using mired in the UI seems unnecessarily technical.  I've worked in the lighting industry designing light fixtures and I've never heard of mired until Home Assistant.

**Comment by KapJI** [comment]
Using kelvins makes a lot of sense to me. I needed to google what mired is.

**Comment by janiversen** [comment]
👍 for the use of kelvin in the UI.

**Comment by KapJI** [comment]
Let's reopen this.

**Comment by thecode** [comment]
Another option is to let the device set the preferred unit similar to what we do in color modes.
If I have to choose one it would also be kelvin, but the optimal solution is to support both, the main problem with kelvin to mired conversion is that it is not an accurate conversion so that if the device work in kelvin and you convert to mired you get a value that when converting back will have a small offset.

**Reply by KapJI** [reply]
Yes, I think we shouldn't break compatibility and should keep support of mired but allow integrations to use kelvins as preferred unit and show this in the UI.

**Comment by icemanch** [comment]
Reading the original post I realized several integrations are converting mired to kelvin to send to devices.  There also several integrations that use the mired.  Is there a way to keep the back end as is and just make the front end display kelvin?  Perhaps store both values and let the integrations get the one they need thus preserving the code currently implemented?

**Reply by thecode** [reply]
@icemanch I think the best option is to make the integration declare the preferred  display option so that the integration will only implement the native format the device supports. A helper in the base light component can be used so that if a service call is made with a format the integration does not support the helper will convert it. 
For example:
An integration for a device with native kelvin control declares `SUPPORTS_KELVIN` (or similar) and implements option to set kelvin temperature in `async_turn_on`. If a service call is made using mired the base light component will first translate it to kelvin and call the service, so that the native format will work in the optimal way and the non-native will still work.

This implementation is backward compatible, so that if a user was using a service with mired it will still work. It will work as usual since conversion happen anyhow before, just now it happens in the base light component and not in the integration.
As for the frontend, In my option we should display the color temp using the format supplied by the integration, another option is to make the frontend display format (kelvin/mired) configurable same as we configure display units (metric/imperial) or make the entity start with the native format supplied by the intergraion and allow users to override it same as they do for other properties (such as overriding the icon)

**Reply by icemanch** [reply]
That makes sense on the helper.

I completely disagree on defaulting to the integrations format.  I would challenge you to find any device or UI other then home assistant that presents mired over CCT.  It's a very obscure way to do things and I have not been able to figure out why it's used internally in some devices other then it's a smaller number.  

Having the ability to choose which to display is probably a great compromise.

**Reply by thecode** [reply]
After reading other comments here and checking different specs, I agree the default should be Kelvin. so we should only consider if we want this to be configurable, I can easily say that I don't know anyone who would prefer mired or anyone not using HA even know what it is.

**Reply by KapJI** [reply]
I think we should use kelvins everywhere what is user facing, e.g. UI and configs. Mired can be kept internally for APIs like Hue.

**Reply by james10952001** [reply]
I can only speculate, but I suspect Philips used mired internally is because it's a smaller number, the app does not present color temperature numerically to the user at all though, it is purely visual, at least the Hue Bluetooth app I have. When you're sending data potentially very frequently to multiple devices via low bandwidth RF in a crowded spectrum and processing it on small low power microcontrollers within the bulbs themselves there is good motivation to reduce the number of bits being used to represent the value. When you are presenting a value to the end user via a modern device there is no reason (that I can think of) not to represent it in a standard unit. Especially when we have a unit to express color temperature that is as far as I know, universally accepted around the globe, certainly throughout the lighting industry.

**Reply by thecode** [reply]
I believe they used it because at that time they thought this would be the correct way, in terms of bits, the packet overhead in zigbee/wifi is so big and MCUs which can accept it won't matter by 8 or 16 bit number. For Bluetooth we can argue since overhead is smaller, anyhow this discussion is not relevant as most people here prefer Kelvin.
A nice article about why mired is used:
https://visualsproducer.wordpress.com/2020/11/29/mireds-versus-degrees-kelvin-for-colour-temperature/

**Comment by james10952001** [comment]
How it works in the back end is irrelevant to the user, I don't think there's anything wrong with how that part is already. The UI should definitely display in kelvin though. Everyone in the lighting industry uses kelvin, color temperature on lamp packages is always specified in kelvin, it makes no sense to present the user with a totally nonstandard esoteric unit like mired even when there are totally valid reasons for using it internally. Displaying color temperature in mired (with no option to change it) makes as much sense as displaying room temperature in kelvin, or distance in furlongs or displaying everything in two's complement binary.

**Reply by thecode** [reply]
@james10952001 As far as I know HA started with mired due to Philips Hue supporting only control by mired, looking at specs for new bulbs I see they write the color in Kelivn. However I don't see any reason why the UI should force something.

Metric units are also industry standard, but still for some _unknown_ :-) reason people prefer to use imperial units.
Btw, the frontend is just a bar with two limits so there is no real work there besides setting the correct limits according to user configuration. The explanation of how it can be implemented in the backend is of course irrelevant to the user, but was written to explain that the implementation is simple and backward compatible.

With all that being said, seeing that Philips also spec their bulbs using Kelvin and that it is a de facto standard I support the option to migrate to Kelvin even if it would be a breaking change for some users.

**Reply by amelchio** [reply]
The backend representation is not quite irrelevant to the user since it leaks to the light attributes which affects at least templates and scenes.

**Comment by Floris-88** [comment]
Would love to have the color temp in the UI in Kelvin! Makes a lot more sense. Also, the range is of the colour temperature slider needs to be adjustable, it now doesnt run fully left - right for my Hue bulbs, it automatically jumps a bit back towards the center.

**Reply by icemanch** [reply]
The range is set by the integration.  You can put an issue in on the Hue integration to get it fixed.

**Comment by mkammes** [comment]
Add me to the chorus.

I'd love to be able to input and display light color temperatures in Kelvin (K) instead of mireds.

Perhaps a global toggle from mireds to Kelvin or the option to display the Kelvin values on a (light) card-by-card basis.

**Comment by ricarva** [comment]
Just another user chiming in: an option to set the default to Kelvin (in the frontend) would be most welcome.

Mireds are not common parlance in the lighting ecosystem these days.

**Comment by cinghialino** [comment]
let's get this going! who uses mired anyways? 

_do-it!, do-it!_

**Comment by GizmoGoody** [comment]
I just created my first scene in HA after using ST for years. I went to set the color temperature of some lights and I thought I messed something up since the values were in the 100's vs. 1000's. Using Kelvin in the UI seems like a no-brainer if you want to lower barriers to entry and welcome less technical users to the platform.

After giving this (https://visualsproducer.wordpress.com/2020/11/29/mireds-versus-degrees-kelvin-for-colour-temperature/#:~:text=The%20formula%20to%20go%20from,2700K%20equals%20370%20mired%2C%20etc.) a quick read it seems the simple implementation would be to change the UI frontend to a Kelvin scale and then set the mired on the backend to 1,000,000/the UI K (e.g. 3000K = 333mired).

**Reply by ristomatti** [reply]
I'm under the impression the conversion logic is already built. So I doubt it's about figuring the algorithm but more on how to handle all the various integrations, UI components and also taking into account existing users, community plugins. That said, I completely agree with everything else. I've had all my lighting based on LIFX bulbs/LED strips for many years.

LIFX products specs define the supported kelvin range and by default their app shows a "color wheel" that's not stepless like the color wheel. It has specific kelvin values like 2500, 2700, 3000, 3250, 3500 etc. I've found this great UX as it's easier to match different lights when the values aren't specific. These numbers are also what you find on specifications for various lights.

I haven't noticed anyone else mention this but I also feel the color temperature range is reverse. I guess this is related to the mired scale to be reverse of kelvin scale, so higher mired value is a warmer color. It makes sense to have the value to increase from left to right but to a person thinking in kelvins it just feels reversed. :slightly_smiling_face:

**Reply by GizmoGoody** [reply]
Mostly agree, except most users don't care about what's going on in the backend...we really only need it to adjust at the UX (1,000,000/mired) or (1,000,000/K) depending on which way you're going. In other words, if all the integrations are currently built using mireds, don't change anything on the backend...just change the frontend to accept K from the user and output mired to the backend (and vice versa for changes made external to that session).

I also noticed your point about the spectrum being reversed (i.e. it's bad UX for at least two reasons ;-))

**Reply by amelchio** [reply]
This is the "architecture" area which is indeed about backend stuff, for example the `color_temp` attribute on lights.

I do think it makes sense to do just the UI part as a start. Just converting the tooltip value would go a long way. We already do that for the brightness (which is 0-255 internally). But you should probably start a discussion about that in the "frontend" area: https://github.com/home-assistant/frontend/discussions

**Reply by GizmoGoody** [reply]
Great point...sorry about that. Here's the frontend discussion and poll:

https://github.com/home-assistant/frontend/discussions/12599

Feel free to cast your vote!

**Reply by devedsmith** [reply]
Voted for Kelvin!!

**Reply by spencerthayer** [reply]
The spectrum being reversed annoys me to no end.

**Comment by spencerthayer** [comment]
Why hasn't this been implemented as an option? This seems to be a much requested feature and it seems objectively simple to have the front end display Kelvins.

**Comment by cinghialino** [comment]
let's do it

**Comment by emontnemery** [comment]
# Updated proposal 2022

* Rename `ATTR_KELVIN` to `ATTR_COLOR_TEMP_KELVIN`
* We currently have `ATTR_KELVIN` as an alternative way to provide `ATTR_COLOR_TEMP`. Reverse this relationship, making `ATTR_COLOR_TEMP_KELVIN` the primary attribute that light platforms receive in their `turn_on` handler.
* Introduce a `LightEntity.color_temp_kelvin` property for the color temperature of a light. Keep support for lights implementing `LightEntity.color_temp` during a deprecation period, just like `hs_color` auto-converts to `rgb_color`.
* Introduce `LightEntity.min_color_temp_kelvin` / `LightEntity.max_color_temp_kelvin` attributes. Keep support for lights implementing `LightEntity.min_mireds` / `LightEntity.max_mireds` during a deprecation period.
* Introduce state and capability attributes `ATTR_COLOR_TEMP_KELVIN`, `ATTR_MIN_COLOR_TEMP_KELVIN` and `ATTR_MAX_COLOR_TEMP_KELVIN`
* Change the UI to use a kelvin scale.
* Deprecate constants `ATTR_KELVIN` and `ATTR_COLOR_TEMP` from the `light.turn_on` service call
* Deprecate state and capability attributes `ATTR_COLOR_TEMP`, `ATTR_MIN_MIREDS` and `ATTR_MAX_MIREDS`
* Deprecate properties `LightEntity.color_temp`, `LightEntity.min_mireds` and `LightEntity.max_mireds`

**Reply by gyptiangodish-prog** [reply]
Aw

**Reply by gyptiangodish-prog** [reply]
I’m reds

**Comment by sebastianrasor** [comment]
Has there been any movement on this? It's pretty strange that the frontend shows the temperature in kelvin, then it's converted to mireds, then converted *back* to kelvin before being sent to the device-specific API or protocol. It's loss of precision with no purpose or function.

The maximum temperature of my device is 6500 kelvin, but I can only set it as high as 6493 kelvin due to unit conversion.
![Screenshot from 2023-03-20 20-23-45](https://user-images.githubusercontent.com/92653912/226519145-5e2057ea-e184-462b-a6ab-a967a98ff33f.png)

The same issue occurs for the lower bounds. The minimum for this device is 2700, however I can only set it as low as 2702 kelvin.
![Screenshot from 2023-03-20 20-24-16](https://user-images.githubusercontent.com/92653912/226520085-9aff92fd-e0f2-4434-8c67-433c3c6399c0.png)

In addition, there's a frustrating visual effect when setting the temperature. 
[Screencast from 03-20-2023 08:24:30 PM.webm](https://user-images.githubusercontent.com/92653912/226520472-c3c00698-6bfe-4e76-8acc-7937f6941631.webm)

**Reply by rytilahti** [reply]
I'm not sure what kind of movement are you looking for? [Kelvins have been the preferred unit for some time](https://github.com/home-assistant/core/pull/79591), but it's up to individual integrations to handle that. If you want to contribute by converting an existing integration to use kelvins, maybe [this conversion PR](https://github.com/home-assistant/core/pull/85646) can be a helpful reference.

**Reply by sebastianrasor** [reply]
It seems like my problems were with ESPHome, not HA

**Comment by m-khvoinitsky** [comment]
I'm going to defend mireds here. The big advantage of mireds is that their perceived difference is linear, e.g. if you make an automation which does +/-25% on button press, you will perceive roughly +/-25% of yellowness/blueness on each button press. As opposed to Kelvins where the same +/-25% would give you something like a perceived +50% on the first press, then +25% and so on.

**Comment by muddy79** [comment]
I'm late to the party, but... honestly... after being forced to review dozens of my old automations, ALWAYS using color_temp, I'd like to rant here a lot. 
WHY mired scale and color_temp has been removed? The calculation is straitforward. 
Are you also planning to remove RGB and HUE/SATURATION and just leave color_xy attribute? If not, why we can have a conversion from different color scales, but not color temperature scale? It's insane!

**Comment by roelanddewindt** [comment]
The Kelvin migration that started in this discussion was completed in 2026.3 by @epenet  — clean, well-executed, long overdue. Thank you.

The cleanup exposed an adjacent gap that I'd like to raise here, since this is where the light color temperature architecture is discussed.

**The remaining problem**

Having the correct Kelvin value in `light.turn_on` is necessary but not sufficient for RGB LED strips. HA still converts that Kelvin value to RGB using a formula that assumes standard LED channel primaries. Real strips — particularly inexpensive ones — deviate significantly from those assumptions. The result is that 2700K "warm white" arrives as a correct Kelvin value, gets converted to RGB by a formula that doesn't know what this specific strip's channels actually look like, and produces a visible pink, magenta, or greenish cast. Alternatively, one has to provide RGB values themselves.

There is currently no mechanism to correct for this. The only workaround is a template light in YAML that intercepts the service call and substitutes hand-tuned RGB values — undiscoverable, requires significant YAML expertise, and provides no UI.

The same problem affects Adaptive Lighting (see [adaptive-lighting#421](https://github.com/basnijholt/adaptive-lighting/discussions/421)), which generates correct Kelvin values that get mangled by the same conversion step.

**Proposed solution**

An optional per-entity calibration profile: the user provides empirically determined anchor points mapping Kelvin values to the RGB output that actually looks correct on their specific strip. HA interpolates between anchors for any requested color temperature, substituting the corrected RGB values instead of using the generic formula. Entities without a profile behave identically to today — fully backwards compatible.

The minimum viable form is a two-anchor linear profile in `configuration.yaml`. A more complete implementation could support a multi-point lookup table and eventually a full 3×3 color matrix (analogous to an ICC profile) for accurate hue correction across the color wheel.

**Working proof-of-concept**

To make this concrete rather than theoretical, I built a working custom component that implements the minimum viable case. 

- Feature request / discussion: https://github.com/orgs/home-assistant/discussions/3538
- PoC repo: https://github.com/roelanddewindt/ha-calibrated_led

The calibration logic is a single self-contained file (`color_profile.py`) with no HA dependency — 13 passing unit tests, linear interpolation between anchor points, linear extrapolation beyond them. The hook point in `light.turn_on` is ~10 lines. Tested on HA 2026.3 with a Zigbee RGB strip.

Tagging @epenet  and @emontnemery  as the authors of the Kelvin migration work — this feels like a natural next step in the same direction, and I'd very much welcome your thoughts on where a calibration profile hook would best live in the light platform. Mind you, I am not a developer, I merely know how to coach AI to put my reasoning into the code provided. I had to learn how to use GitHub to even get this repo online. My wish is to get the attention necessary for some real coding people to include the demonstrated capability into the core of HA.
